lib/PlatoonService/src/V2VCommManager.cpp

```diff
@@ -0,0 +1,534 @@
+ *****************************************************************************/
+
+/* MQTT subtopic name for waypoint reception. */
+const char* V2VCommManager::TOPIC_NAME_WAYPOINT_RX = "inputWaypoint";
+
+/* MQTT subtopic name for platoon heartbeat. */
+const char* V2VCommManager::TOPIC_NAME_PLATOON_HEARTBEAT = "heartbeat";
+
+/* MQTT subtopic name for platoon heartbeat. */
+const char* V2VCommManager::TOPIC_NAME_PLATOON_HEARTBEAT_RESPONSE = "heartbeatResponse";
+
+/** Buffer size for JSON serialization of heartbeat messages. */
+static const uint32_t JSON_HEARTBEAT_MAX_SIZE = 128U;
+
+/******************************************************************************
+ * Public Methods
+ *****************************************************************************/
+
+V2VCommManager::V2VCommManager(MqttClient& mqttClient) :
+    m_mqttClient(mqttClient),
+    m_waypointQueue(),
+    m_waypointInputTopic(),
+    m_waypointOutputTopic(),
+    m_platoonHeartbeatTopic(),
+    m_heartbeatResponseTopic(),
+    m_participantType(PARTICIPANT_TYPE_UNKNOWN),
+    m_platoonId(0U),
+    m_vehicleId(0U),
+    m_lastPlatoonHeartbeatTimestamp(0U),
+    m_followerResponseCounter(0U),
+    m_platoonHeartbeatTimer(),
+    m_vehicleHeartbeatTimeoutTimer(),
+    m_v2vStatus(V2V_STATUS_NOT_INIT),
+    m_vehicleStatus(),
+    m_followers{}
+{
+}
+
+V2VCommManager::~V2VCommManager()
+{
+}
+
+bool V2VCommManager::init(uint8_t platoonId, uint8_t vehicleId)
+{
+    bool isSuccessful = false;
+
+    m_vehicleId = vehicleId;
+
+    if (NUMBER_OF_FOLLOWERS < m_vehicleId)
+    {
+        /* Invalid ID. */
+        LOG_ERROR("Invalid vehicle ID: %d. Maximum followers: %d.", m_vehicleId, NUMBER_OF_FOLLOWERS);
+    }
+    else if (PLATOON_LEADER_ID == m_vehicleId)
+    {
+        /* Its the leader. */
+        m_participantType = PARTICIPANT_TYPE_LEADER;
+    }
+    else
+    {
+        /* Its a follower. */
+        m_participantType = PARTICIPANT_TYPE_FOLLOWER;
+    }
+
+    if (PARTICIPANT_TYPE_UNKNOWN == m_participantType)
+    {
+        LOG_ERROR("Failed to determine participant type.");
+    }
+    else if (false == setupWaypointTopics(platoonId, m_vehicleId))
+    {
+        LOG_ERROR("Failed to setup waypoint topics.");
+    }
+    else if (false == setupHeartbeatTopics(platoonId, m_vehicleId))
+    {
+        LOG_ERROR("Failed to setup heartbeat topics.");
+    }
+    else
+    {
+        isSuccessful = true;
+    }
+
+    if ((true == isSuccessful) && (PARTICIPANT_TYPE_LEADER == m_participantType))
+    {
+        isSuccessful = setupLeaderTopics();
+        m_platoonHeartbeatTimer.start(PLATOON_HEARTBEAT_TIMER_INTERVAL);
+    }
+
+    if (true == isSuccessful)
+    {
+        m_v2vStatus = V2V_STATUS_OK;
+    }
+
+    return isSuccessful;
+}
+
+V2VCommManager::V2VStatus V2VCommManager::process(VehicleStatus status)
+{
+    m_vehicleStatus = status;
+
+    /* Is MQTT client connected? */
+    if ((true == m_mqttClient.isConnected()) && (V2V_STATUS_OK == m_v2vStatus))

```

Comment:
A v2v status other than OK will never heal.

Reply:
Is intended until an Error handler in the application is developed